chromium_src/chrome/browser/ui/cocoa/first_run_dialog_controller.mm

```diff
@@ -3,15 +3,208 @@
 
+#include "chrome/browser/ui/cocoa/first_run_dialog_controller.h"
+
+#include "base/i18n/rtl.h"
+#include "base/mac/scoped_nsobject.h"
+#include "base/strings/sys_string_conversions.h"
 #include "brave/grit/brave_generated_resources.h"
+#include "chrome/browser/ui/cocoa/key_equivalent_constants.h"
+#include "chrome/grit/chromium_strings.h"
 #include "chrome/grit/generated_resources.h"
+#import "third_party/google_toolbox_for_mac/src/AppKit/GTMUILocalizerAndLayoutTweaker.h"
+#include "ui/base/cocoa/controls/button_utils.h"
+#include "ui/base/cocoa/controls/textfield_utils.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "ui/base/l10n/l10n_util_mac.h"
+
+namespace {
+
+// Return the internationalized message |message_id|, with the product name
+// substituted in for $1.
+NSString* NSStringWithProductName(int message_id) {
+  return l10n_util::GetNSStringF(message_id,
+                                 l10n_util::GetStringUTF16(IDS_PRODUCT_NAME));
+}
+
+// Reflows buttons. Requires them to be passed in vertical order, top down.
+CGFloat VerticallyReflowButtons(NSArray<NSButton*>* buttons) {
+  CGFloat localVerticalShift = 0;
+  for (NSButton* button in buttons) {
+    [GTMUILocalizerAndLayoutTweaker wrapButtonTitleForWidth:button];
+
+    NSRect oldFrame = button.frame;
+    [button sizeToFit];
+    NSRect newFrame = button.frame;
+    // -[NSControl sizeToFit], with no layout constraints, like here, will end
+    // up horizontally resizing and keeping the upper left corner in place.
+    // That wrecks RTL, and because all that's really needed is vertical
+    // resizing, reset the horizontal size.
+    newFrame.size.width = NSWidth(oldFrame);
+    button.frame = newFrame;
+
+    localVerticalShift += NSHeight(newFrame) - NSHeight(oldFrame);
+    if (localVerticalShift) {
+      NSPoint origin = button.frame.origin;
+      origin.y -= localVerticalShift;
+      [button setFrameOrigin:origin];
+    }
+  }
+  return localVerticalShift;
+}
+
+void MoveViewsVertically(NSArray* views, CGFloat distance) {
+  for (NSView* view : views) {
+    NSRect frame = view.frame;
+    frame.origin.y += distance;
+    [view setFrame:frame];
+  }
+}
+
+// Center |view| vertically within its own superview, so its horizontal
+// centerline is the same as its superview's horizontal centerline.
+void CenterVertically(NSView* view) {
+  NSView* superview = view.superview;
+  NSRect frame = view.frame;
+  NSRect superframe = superview.frame;
+  frame.origin.y = (NSHeight(superframe) - NSHeight(frame)) / 2.0;
+  [view setFrame:frame];
+}
+
+}  // namespace
+
+@implementation FirstRunDialogViewController {
+  // These are owned by the NSView hierarchy:
+  NSButton* _defaultBrowserCheckbox;
+}
+
+- (instancetype)initWithStatsCheckboxInitiallyChecked:(BOOL)checked
+                        defaultBrowserCheckboxVisible:(BOOL)visible {
+  // We always make default checkbox visible.
+  DCHECK(visible);
+  return [super init];
+}
+
+- (void)loadView {
```

Comment:
Is there no way to override this and then adjust size/positions so that we don't duplicate all the code?

Reply:
yes, unfortunately.
Subclass can't access vars declared in @implementation section in super class.
Because of that, most of all super class method can't be reusable.

Reply:
I'm not sure which is more better (replacing whole impls vs. subclassing) when most of super class impls can't be reused. WDYT? If you think subclassing is better, I'll do subclassing. (

Reply:
If we can't reuse then there's little point in subclassing, though I wish we could reuse all the functions in the anonymous namespace at least. Maybe @bridiver has some suggestions?

Reply:
Hmm, not sure it's possible to re-use functions in anon-namespace..

Reply:
I'll unblock, if a better solution comes along we can always revisit.